documentation/Semantics.md

```diff
@@ -1,206 +1,139 @@
-      - Block has a specification part
-      - Blocks start Step 1..4 again
-      - N.B. Implicits are applied to the host scope
-    * Implied Do
-    * Index names in Forall and Do Concurrent
-    * Change Team
-    * OpenMP and OpenACC constructs
-    * ENTRY
-
-References to derived types members are not resolved until semantics
-
-No semantic checking or resolving of types (except for implicit declarations)
-has happened yet.
-
-#### Step 5. Perform Step 1..4 on each internal procedure
-- Side effect is that each internal procedure gets a proper interface in the
-parent scope
-- We do this now because we need to know the return and argument types for
-functions, e.g. `a = f(a, b, c) % x + 1`
-
-#### Step 6. Tree Disambiguation
-
-At this point, or during Step 3 (TBD), the tree can be rewritten to be
-unambiguous.
-- Structure vs operator a.b.c.d
-- Array references vs function calls
-- Statement functions vs array assignment (In Step 3)
-- READ/WRITE stmts where the arguments do not have keywords
-  - WRITE (6, X)  ….
-  - That X might be a namelist group or an internal variable
-  - Need to know the names of the namelist groups to disambiguate it
-- Others….? TBD
-
-Resolution of parse tree ambiguity (statement function definition, function vs.
-  array)
-
-#### Step 7. Do enough semantic processing to generate .mod files
-- Fully resolve derived types
-- Combine and check declarations of all entities within a given scope; resolve
-their type, rank, shape, and other attributes.
-- Constant evaluation is required at this point.
-
-Why do Step 7 before the rest of semantic checking? The sooner we can generate
-mod file the sooner we can read ‘em; you can test a lot of Fortran programs as
-soon as you can read mod files.
-
-#### Step 8. Semantic Rule Checking
-
-An incomplete and unordered list of requirements for semantic analysis:
-
-* EQUIVALENCE overlaying (checking at least)
-* Intrinsic function generic->specific resolution, constraint checking, T/R/S.
-* Compile-time evaluation of constant expressions, including intrinsic
-functions.
-* Resolution of generics and type-bound procedures.
-* Identifying and recording uplevel references.
-* Control flow constraint checking
-* Labeled DO loop terminal statement expansion? (maybe not, can defer to CFG in
-  DST).
-* Construct association: distinguish pointer-like from allocatable-like
-* OMP and OACC checking
-* CUF constraint checking
-
-## Utility Routines
-
-### Diagnostic Output
-TBD
-
-### Constant Expression Evaluation
-- Scalars
-- Array intrinsics
+If the program is legal, the semantic pass will produce a (possibly modified)
+parse tree for the semantically correct program with each name mapped to a symbol
+and each expression fully analyzed.
+
+All user errors are detected during semantic analysis.
+After it completes successfully the program should compile with no error messages.
+There may still be warnings or informational messages.
+
+## Phases of Semantic Analysis
```

Comment:
It would be good to know what happens in the compiler before semantic analysis -- options are processed (some of which affect semantic analysis), a "cooked" character stream is constructed, the program is parsed, and source provenance information is generated.

Reply:
This is in Overview.md. Do you think we should mention that here?

Reply:
I just looked at Overview.md.  There's a lot of information that's duplicated in this document.

I do think that it's good for the reader of this document to have an idea of what the inputs are to semantic analysis even if that repeats some information.  Also, it looks to me like Overview.md should be rewritten once this document is updated.  For example, both documents describe label validation and name resolution.

Reply:
I agree that Overview.md will need some work.

I'll add a little more to this document about the inputs.